src/core/schema.js

```diff
@@ -0,0 +1,117 @@
+var utils = require('../utils/');
+
+var coordinates = utils.coordinates;
+var types = module.exports.types = {};
+
+registerType('boolean', booleanParse, booleanStringify);
+registerType('coordinate', coordinates.parse, coordinates.stringify);
+
+/**
+ * A build step to the schema to expand `type` into built-in parsers and stringifiers.
+ */
+module.exports.process = function (schema) {
+  Object.keys(schema).forEach(function (property) {
+    var schemaPropDefinition = schema[property];
+    var type = schemaPropDefinition.type;
+    var typeParser;
+
+    if (type) {
+      typeParser = types[type];
+      schemaPropDefinition.parse = typeParser.parse;
+      schemaPropDefinition.stringify = typeParser.stringify;
+    }
+  });
+  return schema;
+};
+
+/**
+ * If `value` is object, coerce values of `val` into types defined by `schema`.
+ *
+ * If `value` is string and `schemaPropDefinition` defined, coerces `value` into the value
+ * in `schema` pointed to by `schemaPropDefinition`.
+ *
+ * In case `value` is a string, `schemaPropDefinition` signifies which value in `schema` to
+ * coerce to.
+ *
+ * @param {object|string} value - value(s) to coerce.
+ * @param {object} schema - Object which values will be used to coerce to.
+ * @param {string} schemaPropDefinition
+ * @returns Coerced value or object.
+ */
+module.exports.coerce = function (value, schema, schemaProp) {
+  var obj = value;
+  var schemaPropDefinition;
+
+  // Batch coerce.
+  if (value.constructor === Object) {
+    Object.keys(obj).forEach(function (key) {
+      var schemaPropDefinition = schema[key];
+      var schemaValue;
+      if (!schemaPropDefinition) {
+        utils.warn('Unknown component property: ' + key);
+        return;
+      }
+      schemaValue = schemaPropDefinition.default;
+      if (schemaValue === undefined) { return; }
+      obj[key] = coerceValue(obj[key], schemaValue, schemaPropDefinition.parse);
+    });
+    return obj;
+  }
+
+  // Single-value coerce.
+  schemaPropDefinition = schema[schemaProp];
+  if (!schemaPropDefinition) {
+    utils.warn('Unknown component property: ' + schemaProp);
+    return;
+  }
+  return coerceValue(value, schemaPropDefinition.default, schemaPropDefinition.parse);
+};
+
+function coerceValue (value, targetValue, parser) {
+  if (typeof value !== 'string') { return value; }
+
+  if (parser && parser.constructor === Function) {
```

Comment:
typeof parser === 'function'?

http://stackoverflow.com/questions/5999998/how-can-i-check-if-a-javascript-variable-is-function-type